Introduce salt into row hash (#3528) - only part 2

yoniko opened the pull request:
This helps to avoid regressions where consecutive compressions use the same tag space with similar data (running `zstd -b5e7 enwik8 -B128K` reproduces this regression).

yoniko committed:
Add init once memory
- Adds memory type that is guaranteed to have been initialized at least once in the workspace's lifetime.
- Changes tag space in row hash to be based on init once memory.
- Moves buffers to aligned memory and removes the buffer memory type.

yoniko committed:
Restores unaligned buffer allocations, but changes allocation order to allocate them after aligned buffers

yoniko committed:
Introduce salt into row hash (#3528)
This helps to avoid regressions where consecutive compressions use the same tag space with similar data (running `zstd -b5e7 enwik8 -B128K` reproduces this regression).

Cyan4973 commented on lib/compress/zstd_compress_internal.h:
```diff
@@ -828,6 +836,24 @@ size_t ZSTD_hashPtr(const void* p, U32 hBits, U32 mls)
+    switch(mls)
+    {
+        default:
+        case 4: return ZSTD_hash4PtrS(p, hBits, (U32)hashSalt);
```
I would rather recommend to use the upper part of `hashSalt` (`(U32)(hashSalt >> 32)`),
as it's less predictable than the lower part.

An alternative approach could be to make `hashSalt` evolution of lower bits less predictable.

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -1936,6 +1936,13 @@ ZSTD_reset_matchState(ZSTD_matchState_t* ms,
+        /* We want to generate a new salt in case we reset a Cctx, but we always want to use
+         * 0 when we reset a Cdict */
+        if(forWho == ZSTD_resetTarget_CCtx) {
+            ms->hashSalt = ms->hashSalt * 6364136223846793005 + 1; /* based on MUSL rand */
```
There is probably enough time here to make `hashSalt` evolution more chaotic, closer to a prng.

yoniko committed:
CR fixes

yoniko committed:
CR fixes

yoniko committed:
CR fixes and better entropy for hash salt

yoniko committed:
Slightly improve entropy performance

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -1935,7 +1935,18 @@ ZSTD_reset_matchState(ZSTD_matchState_t* ms,
+         * 0 when we reset a Cdict */
+        if(forWho == ZSTD_resetTarget_CCtx) {
+            ms->tagTable = (U16 *) ZSTD_cwksp_reserve_aligned_init_once(ws, tagTableSize);
+            ms->hashSalt = (U64) ZSTD_hashPtr(&ms->hashSalt, 32, sizeof(ms->hashSalt)) << 32 |
```
I would rather go for a stronger mixing, making observation of any portion of bits essentially random (including lower bits).

I would recommend to base it on  `XXH3_rrmxmx()` :
https://github.com/Cyan4973/xxHash/blob/v0.8.1/xxhash.h#L3403

This mixing has been tested and proved strong enough to pass `practrand` tests,
making it essentially as strong as a PCG random number generator.

Cyan4973 commented on lib/common/bits.h:
```diff
@@ -172,4 +172,29 @@ MEM_STATIC unsigned ZSTD_highbit32(U32 val)   /* compress, dictBuilder, decodeCo
+    return (value >> count) | (U64)(value << ((0U - count) & 0x3F));
+}
+
+FORCE_INLINE_TEMPLATE
```
minor: do you need all 3 variants ?
Only `ZSTD_rotateRight_U64()` seems in use.

yoniko replied:
All 3 variants are currently in use in `ZSTD_row_matchMaskGroupWidth`, I opted to moving all of them here from `zstd_lazy.c` instead of just one as it's the more natural place for them in any case.

yoniko committed:
Merge branch 'init-once-memory-3528-part1' into tag-space-hash-salting-3528-part2

yoniko committed:
Merge bugfix

yoniko committed:
Use `MEM_STATIC` instead of `FORCE_INLINE_TEMPLATE`

yoniko committed:
Add init once memory (#3528) (#3529)

- Adds memory type that is guaranteed to have been initialized at least once in the workspace's lifetime.
- Changes tag space in row hash to be based on init once memory.

yoniko committed:
Merge remote-tracking branch 'origin/dev' into tag-space-hash-salting-3528-part2

yoniko committed:
merge fix
